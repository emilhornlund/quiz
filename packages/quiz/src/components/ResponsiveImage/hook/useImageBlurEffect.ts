import { CountdownEvent } from '@quiz/common'
import { useEffect, useRef, useState } from 'react'

export const useImageBlurEffect = (countdown?: CountdownEvent) => {
  const [blur, setBlur] = useState(5)
  const [clientToServerOffset, setClientToServerOffset] = useState(0)
  const [initiatedTime, setInitiatedTime] = useState(0)
  const [totalDuration, setTotalDuration] = useState(0)
  const intervalRef = useRef<ReturnType<typeof setInterval> | null>(null)

  useEffect(() => {
    if (intervalRef.current) {
      clearInterval(intervalRef.current)
      intervalRef.current = null
    }
    if (!countdown) return

    const serverTime = new Date(countdown.serverTime).getTime()
    const tmpInitiatedTime = new Date(countdown.initiatedTime).getTime()
    const expiryTime = new Date(countdown.expiryTime).getTime()

    const offset = serverTime - Date.now()
    setClientToServerOffset(offset)

    const duration = Math.max((expiryTime - tmpInitiatedTime) / 1000, 1)
    setTotalDuration(duration)
    setInitiatedTime(tmpInitiatedTime)

    const now = Date.now() + offset
    const elapsed = (now - tmpInitiatedTime) / 1000
    const initialBlur = Math.max(((duration - elapsed) / duration) * 5, 0)
    setBlur(initialBlur)
  }, [countdown])

  useEffect(() => {
    if (!initiatedTime || !totalDuration || intervalRef.current) return

    intervalRef.current = setInterval(() => {
      const now = Date.now() + clientToServerOffset
      const elapsed = (now - initiatedTime) / 1000

      if (elapsed >= totalDuration) {
        setBlur(0)
        clearInterval(intervalRef.current!)
        intervalRef.current = null
        return
      }

      const newBlur = Math.max(
        ((totalDuration - elapsed) / totalDuration) * 5,
        0,
      )
      setBlur(newBlur)
    }, 100)

    return () => {
      if (intervalRef.current) clearInterval(intervalRef.current)
    }
  }, [clientToServerOffset, initiatedTime, totalDuration])

  return {
    filter: `blur(${blur.toFixed(2)}rem)`,
    clipPath: 'inset(2px)',
  }
}
