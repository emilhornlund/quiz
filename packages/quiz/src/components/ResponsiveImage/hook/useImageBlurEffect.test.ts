import { renderHook } from '@testing-library/react'
import { describe, expect, it } from 'vitest'

import { useImageBlurEffect } from './useImageBlurEffect'

const countdown = {
  serverTime: new Date().toISOString(),
  initiatedTime: new Date(Date.now() - 2000).toISOString(),
  expiryTime: new Date(Date.now() + 2000).toISOString(),
}

describe('useImageBlurEffect', () => {
  it('returns default filter and clipPath when no countdown', () => {
    const { result } = renderHook(() => useImageBlurEffect())
    expect(result.current).toMatchSnapshot()
  })

  it('returns filter and clipPath with countdown', () => {
    const { result } = renderHook(() => useImageBlurEffect(countdown))
    expect(result.current).toMatchSnapshot()
  })
})
