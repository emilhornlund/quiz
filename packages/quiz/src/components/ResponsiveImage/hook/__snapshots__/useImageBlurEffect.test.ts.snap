// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useImageBlurEffect > returns default filter and clipPath when no countdown 1`] = `
{
  "clipPath": "inset(2px)",
  "filter": "blur(5.00rem)",
}
`;

exports[`useImageBlurEffect > returns filter and clipPath with countdown 1`] = `
{
  "clipPath": "inset(2px)",
  "filter": "blur(2.50rem)",
}
`;
